pkg/icingadb/runtime_updates.go

```diff
@@ -95,57 +96,198 @@ func (r *RuntimeUpdates) Sync(
-			sem := semaphore.NewWeighted(1)
+				// Updates must be executed in order, ensure this by using a semaphore with maximum 1.
+				sem := semaphore.NewWeighted(1)
 
-			onSuccess := []OnSuccess[contracts.Entity]{
-				OnSuccessIncrement[contracts.Entity](&counter), OnSuccessIncrement[contracts.Entity](stat),
+				onSuccess := []OnSuccess[contracts.Entity]{
+					OnSuccessIncrement[contracts.Entity](&counter), OnSuccessIncrement[contracts.Entity](stat),
+				}
+				if !allowParallel {
+					onSuccess = append(onSuccess, OnSuccessSendTo(upsertedFifo))
+				}
+
+				return r.db.NamedBulkExec(
+					ctx, upsertStmt, upsertCount, sem, entities, com.SplitOnDupId[contracts.Entity], onSuccess...,
+				)
 			}
-			if !allowParallel {
-				onSuccess = append(onSuccess, OnSuccessSendTo(upsertedFifo))
+		}
+
+		deleteEntityFunc := func(deleteIds <-chan interface{}) func() error {
+			return func() error {
+				var counter com.Counter
+				defer periodic.Start(ctx, r.logger.Interval(), func(_ periodic.Tick) {
+					if count := counter.Reset(); count > 0 {
+						r.logger.Infof("Deleted %d %s items", count, s.Name())
+					}
+				}).Stop()
+
+				sem := r.db.GetSemaphoreForTable(utils.TableName(s.Entity()))
+
+				onSuccess := []OnSuccess[any]{OnSuccessIncrement[any](&counter), OnSuccessIncrement[any](stat)}
+				if !allowParallel {
+					onSuccess = append(onSuccess, OnSuccessSendTo(deletedFifo))
+				}
+
+				return r.db.BulkExec(ctx, r.db.BuildDeleteStmt(s.Entity()), deleteCount, sem, deleteIds, onSuccess...)
 			}
+		}
 
-			return r.db.NamedBulkExec(
-				ctx, upsertStmt, upsertCount, sem, upsertEntities, com.SplitOnDupId[contracts.Entity], onSuccess...,
-			)
-		})
+		switch s.Entity().(type) {
+		case *v1.Host, *v1.Service:
+			entities := make(chan contracts.Entity, 1)
+			g.Go(func() error {
+				defer close(entities)
 
-		g.Go(func() error {
-			var counter com.Counter
-			defer periodic.Start(ctx, r.logger.Interval(), func(_ periodic.Tick) {
-				if count := counter.Reset(); count > 0 {
-					r.logger.Infof("Deleted %d %s items", count, s.Name())
+				var counter com.Counter
+				defer periodic.Start(ctx, r.logger.Interval(), func(_ periodic.Tick) {
+					if count := counter.Reset(); count > 0 {
+						r.logger.Infof("Upserted %d %s sla lifecycles", count, s.Name())
+					}
+				}).Stop()
+
+				first, forward, err := com.CopyFirst(ctx, CreateSlaLifecyclesFromCheckables(ctx, g, upsertEntities, false))
+				if first == nil {
+					return errors.Wrap(err, "can't copy first entity")
 				}
-			}).Stop()
 
-			sem := r.db.GetSemaphoreForTable(utils.TableName(s.Entity()))
+				stmt, _ := r.db.BuildInsertIgnoreStmt(first)
 
-			onSuccess := []OnSuccess[any]{OnSuccessIncrement[any](&counter), OnSuccessIncrement[any](stat)}
-			if !allowParallel {
-				onSuccess = append(onSuccess, OnSuccessSendTo(deletedFifo))
-			}
+				// We can only bulk exec one query at a time, otherwise, the regression test won't pass
```

Comment:
> the regression test won't pass

Bonus points for telling why apart form the tests. I mean, in theory, there could also be a problem in the tests.